File: packages/core/src/programmers/iterate/json_schema_tuple.ts
import { OpenApi } from "@typia/interface";

import { MetadataSchema } from "../../schemas/metadata/MetadataSchema";
import { MetadataTuple } from "../../schemas/metadata/MetadataTuple";
import { json_schema_station } from "./json_schema_station";

export const json_schema_tuple = (props: {
  components: OpenApi.IComponents;
  tuple: MetadataTuple;
}): OpenApi.IJsonSchema.ITuple => {
  const tail: MetadataSchema | null =
    props.tuple.type.elements.at(-1)?.rest ?? null;
  const prefixItems: MetadataSchema[] = props.tuple.type.isRest()
    ? props.tuple.type.elements.slice(0, -1)
    : props.tuple.type.elements;
  return {
    type: "array",
    prefixItems: prefixItems.map((metadata) =>
      json_schema_station({
        blockNever: false,
        components: props.components,
        metadata,
        attribute: {},
      }),
    ),
    additionalItems: tail
      ? json_schema_station({
          blockNever: false,
          components: props.components,
          metadata: tail,
          attribute: {},
        })
      : false,
  };
};

File: packages/core/src/programmers/iterate/json_schema_tuple.go
package iterate

import h "github.com/samchon/typia/packages/core/src/programmers/helpers"

func json_schema_tuple(tuple h.Metadata) h.Schema {
	items := make([]h.Schema, 0)
	for _, elem := range h.GetSlice(tuple, "elements") {
		items = append(items, json_schema_station(h.AsMetadata(elem)))
	}
	schema := h.Schema{"type": "array", "prefixItems": items, "minItems": len(items)}
	if postfix := postfix_of_tuple(tuple); postfix == "" {
		schema["maxItems"] = len(items)
	}
	return schema
}
